### PR TITLE
Deprecate `@auto_move_data` in favor of `trainer.predict`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title.

This was added in 0.8.0.

A [quick code search](https://github.com/search?q=%22auto_move_data%22&type=code) shows some users of this feature, but many don't use it correctly or assume it's required. Additionally, we want people to use `predict`.

Also improves predict docs.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Comment by dsantiago** [comment]
I am finding some trouble with `Trainer.predict()`. I am trying this:

```python
#this
trainer.predict(model, target)

# and this:
model.eval()
trainer.predict(model, target)
```

But even with `eval()` i am getting **BatchNorm error** ... in the end as my target dataset is small i am the follow as they work as should:

```python
# this
def predictions(model, inputs):
  model.eval()
  with torch.no_grad():
    preds = model(inputs)
  
  model.train()
  return preds    

predictions(model, test_data)

# or this
model.predict(test_data, batch_idx=1) # as my model is from pl.lightining module, batch_id can be whatever
```

The question is, why `Trainer.predict()` do not use the `model.eval() + torch.no_grad()` approach ?

**Comment by carmocca** [comment]
> I am finding some trouble with `Trainer.predict()`. I am trying this: ...

Hi @dsantiago, can you open a separate issue about this? A reproduction snippet would also be greatly appreciated.

**Comment by dsantiago** [comment]
Hi @carmocca it's just a point, not an error, i am just asking if is it OK for Trainer.predict() run Dropout and BatchNorm layers, because for me it's counter intuitive.... But i can always rely in calling model directly, with model.eval() + torch.nograd()  or use the pytorch lightning model's predict() function... 

It's just to reflect...

**Comment by carmocca** [comment]
> it's just a point, not an error

It should be getting called:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4c07ab5e99dd20c1f309d9e73cdaacc1ebad9499/pytorch_lightning/trainer/trainer.py#L797-L800

So if it is not, it's an error.

So please, try to reproduce it and open an issue about it 😃

**Comment by dsantiago** [comment]
> > it's just a point, not an error
> 
> It should be getting called:
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/4c07ab5e99dd20c1f309d9e73cdaacc1ebad9499/pytorch_lightning/trainer/trainer.py#L797-L800
> 
> So if it is not, it's an error.
> 
> So please, try to reproduce it and open an issue about it 😃

I saw that part of the code when trying to understand, and it really just disable the grad calculation, if  the model isn't  in eval mode, it will continue to run Dropouts and BatchNorm layers...

Just one call to model.eval() before getting the model predictions and put it on model.train() back after should make de deal(like in the predictions() function i did above).

**Comment by carmocca** [comment]
`on_predict_mode_eval` calls `model.eval()`

**Comment by dsantiago** [comment]
> `on_predict_mode_eval` calls `model.eval()`

Well i remove some boilerplate code so you can verify if u wish, dont need to add gpus as the network is kind of small. Maybe i am missing the point from my side

https://colab.research.google.com/drive/1RYltV3rxRhqyBS2GPv9uOFE8ghb79Tg3?usp=sharing

**Comment by awaelchli** [comment]
I know what the problem is with your code @dsantiago, but let's not discuss that here. Please open a discussion (discussion tab on top next to Pull Requests) ping me there.

**Review by awaelchli — APPROVED** [review]
😢 😢 😢

**Comment by dsantiago** [comment]
> I know what the problem is with your code @dsantiago, but let's not discuss that here. Please open a discussion (discussion tab on top next to Pull Requests) ping me there.

Thanks @awaelchli , did it now! Thanks for your time in seeing it!

**Review by tchaton — APPROVED** [review]
LGTM !
